### ISSUE TITLE
More dynamic PT APIs

### ISSUE THREAD

**Issue by dlibenzi** [pytorch/pytorch#31558]
## 🚀 Feature

Implement some of the pytorch APIs taking tensors as inputs, instead of python scalars.

## Motivation

Many of the pytorch APIs takes some of the inputs as python scalars.
Example:

https://pytorch.org/docs/stable/tensors.html#torch.Tensor.narrow

This is OK when the inputs are step constants (either real constants, or command line arguments), but in case these are dynamic, they force model code to perform `item()` calls on tensors.
For pytorch CPU/GPU an `item()` call is relatively cheap, but for backends (like XLA) which like to work within a *Graph World*, they force explicit execution at intermediate graph locations by hence considerably lower performance.
It would be nice if some of the pytorch APIs would take some of the inputs as tensors.
For example, `torch.narrow()` could take `start` (and maybe even `length`) as tensor, which would map to:
```Python
at::Tensor narrow(const at::Tensor& self, const at::Tensor& start, int64_t lenght);
```
For CPU/GPU this can be easily mapped to the existing, static, `narrow()`, while XLA could override that and map it to `DynamicSlice()`.

Similarly, would be nice to have APIs which return the size of tensors, as tensors.
Example:
```Python
at::Tensor dimsize(const at::Tensor& self, int dim);
```

This came up a few times already, like in:

https://github.com/pytorch/xla/issues/1467

In that case a wavenet like model needs to slice tensors based on offsets (`start`) values coming off training.

**Comment by dlibenzi** [comment]
@ailzhang @mruberry

**Comment by mruberry** [comment]
Thanks for filing the issue, @dlibenzi!

**Comment by cpuhrsch** [comment]
This is a big feature that'll require separate discussion.

**Comment by vadimkantorov** [comment]
(If dynamic torch.arange was supported, torch.narrow can (less efficiently) be simluated as torch.arange + torch.gather)

**Comment by ezyang** [comment]
I dunno, this doesn't seem like a "big feature" problem to me. The basic pitch is to add some new overloads to some functions, accepting `Tensor` where `int` was previously accepted, and then adjusting the kernels to understand this situation. This could be done on a case by case basis, and I don't think there's any infra work we need to do ahead of time to make this happen.

**Comment by dlibenzi** [comment]
Absolutely.
The idea was not to clone the whole API making it dynamic, but to target the case-by-case issues as they come up.

**Comment by gchanan** [comment]
I'm a little unclear about something:

> For CPU/GPU this can be easily mapped to the existing, static, narrow() (via a simple item() call), while XLA could override that and map it to DynamicSlice().

For a call in Python that uses `int`s, this is suggesting we use create Tensors out of the ints to map to the Tensor overload?  With our current Tensor overhead this seems like a no-go.

**Comment by dlibenzi** [comment]
No, that is not what I meant.
Today we have code that is like this:
```Python
(a, b) = some_layer(...) # a and b being Tensors
x = torch.narrow(a, b.item(), WINDOW_SIZE)
...
```

Since `torch.narrow()` takes Python scalars.
The code above could be simply:
```Python
(a, b) = some_layer(...)
x = torch.narrow(a, b, WINDOW_SIZE)
...
```

What I meant with the CPU/GPU statement is that, if we add a new overrideable:
```C++
at::Tensor narrow(const at::Tensor& self, const at::Tensor& start, int64_t lenght);
```

The CPU/GPU backend can simply be generated as:
```C++
at::Tensor narrow(const at::Tensor& self, const at::Tensor& start, int64_t lenght) {
  return narrow(self, start.long(), lenght);
}
```

**Comment by gchanan** [comment]
Yep, there are a bunch of options here, and your last case doesn't have performance downsides for CPU/CUDA.  But this does require that people write PyTorch code differently, in a way that seems hard to get across if we just do the one-off cases that @ezyang suggests.

Like, what do you document?  "Prefer passing Tensors to Python numbers so XLA doesn't have to leave GraphWorld.  Look up the documentation for every function to see if it accepts Tensors where you currently pass Python numbers."

It is of course much easier to get the point across if we accept Tensors everywhere (but that is a bigger feature, as mentioned above).

@dlibenzi: do you think this is an accurate summary of the situation, or did I miss something?

**Comment by dlibenzi** [comment]
I think there are two layers. The python one and the C++ one.
For the C++ one, for performance reasons, it probably makes sense to have different APIs?
For the Python level APIs, it would be sufficient for the one-off cases we add, to simply change the documentation from "a python scalar" to "a python scalar or a rank 0 tensor"?
The the python layer can forward to the proper C++ API according to argument type?

**Comment by gchanan** [comment]
Yes that's right, but doesn't really get at my concern of the one-off (i.e. implement specific cases as they come up) approach.

"Prefer using tensors for scalr values, but you need to check whether each function supports them" isn't a great user experience.

**Comment by dlibenzi** [comment]
Aren't people supposed to be looking at the API documentation before using them?

IMHO backward compatibility is more important than having a blanket statement about what one should be using, in all cases.

I think we want to have a path where, to someone having implemented a model using a scalar argument, we can say `if you pass directly that tensor you should be gaining performance`, in the backends where it matters (and pay virtually no penalties in other backends).
More than issuing a `you shall convert all the pytorch API calls to pass them tensors, as this is the new rule`.

**Comment by ezyang** [comment]
I'm with @dlibenzi here, I suspect people are going to keep reflexively writing `b.item()` for the forseeable future, and someone who is trying to remove sync points will go ahead and try to remove it and see if it will work or not. That seems OK, for people who are trying to squeeze out performance. And maybe someday we can nail all of the spots so that they take scalars, and we can generally give the recommendation you suggest. But it doesn't seem like a big deal to start putting in the infra pieces we need to give this rec, one at a time.

**Comment by gchanan** [comment]
We will accept individual PRs that implement this; we don't have broader plans at this time to change the parsing/code-gen etc. to automatically and universally add support.

**Comment by dlibenzi** [comment]
Thanks @gchanan , that makes sense indeed.
We can target APIs as it fits, and provide individual PRs.
